File: prac5-6/algoritmos.cpp
#include <iostream>
#include <vector>
#include <chrono>
#include <algorithm> // reverse
using namespace std;

using Clock = chrono::high_resolution_clock;

// Estructura para métricas
struct Metrics {
    long long comparisons = 0;
    long long swaps = 0;
};

// ======================= BUBBLE SORT =======================
void bubbleSort(vector<int>& arr, Metrics& m) {
    int n = arr.size();
    bool swapped;

    for (int i = 0; i < n - 1; i++) {
        swapped = false;

        for (int j = 0; j < n - i - 1; j++) {
            m.comparisons++;
            if (arr[j] > arr[j + 1]) {
                swap(arr[j], arr[j + 1]);
                m.swaps++;
                swapped = true;
            }
        }

        if (!swapped) break;
    }
}

// ======================= INSERTION SORT =======================
void insertionSort(vector<int>& arr, Metrics& m) {
    int n = arr.size();

    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;

        while (j >= 0) {
            m.comparisons++;
            if (arr[j] > key) {
                arr[j + 1] = arr[j];
                m.swaps++;
                j--;
            } else break;
        }

        arr[j + 1] = key;
    }
}

// ======================= MERGE SORT =======================
void merge(vector<int>& arr, int left, int mid, int right, Metrics& m) {
    int n1 = mid - left + 1;
    int n2 = right - mid;

    vector<int> L(n1), R(n2);

    for (int i = 0; i < n1; i++) L[i] = arr[left + i];
    for (int i = 0; i < n2; i++) R[i] = arr[mid + 1 + i];

    int i = 0, j = 0, k = left;

    while (i < n1 && j < n2) {
        m.comparisons++;
        if (L[i] <= R[j]) {
            arr[k++] = L[i++];
        } else {
            arr[k++] = R[j++];
        }
        m.swaps++;
    }

    while (i < n1) {
        arr[k++] = L[i++];
        m.swaps++;
    }

    while (j < n2) {
        arr[k++] = R[j++];
        m.swaps++;
    }
}

void mergeSortRec(vector<int>& arr, int left, int right, Metrics& m) {
    if (left < right) {
        int mid = (left + right) / 2;
        mergeSortRec(arr, left, mid, m);
        mergeSortRec(arr, mid + 1, right, m);
        merge(arr, left, mid, right, m);
    }
}

void mergeSort(vector<int>& arr, Metrics& m) {
    mergeSortRec(arr, 0, arr.size() - 1, m);
}

// ======================= QUICK SORT =======================
int partition(vector<int>& arr, int low, int high, Metrics& m) {
    int pivot = arr[high];
    int i = low - 1;

    for (int j = low; j < high; j++) {
        m.comparisons++;
        if (arr[j] < pivot) {
            i++;
            swap(arr[i], arr[j]);
            m.swaps++;
        }
    }

    swap(arr[i + 1], arr[high]);
    m.swaps++;
    return i + 1;
}

void quickSortRec(vector<int>& arr, int low, int high, Metrics& m) {
    if (low < high) {
        int pi = partition(arr, low, high, m);
        quickSortRec(arr, low, pi - 1, m);
        quickSortRec(arr, pi + 1, high, m);
    }
}

void quickSort(vector<int>& arr, Metrics& m) {
    quickSortRec(arr, 0, arr.size() - 1, m);
}

// ======================= TEST =======================
void test(vector<int> data,
          void(*func)(vector<int>&, Metrics&),
          const string& name) {

    vector<int> arr = data;
    Metrics m;

    auto t1 = Clock::now();
    func(arr, m);
    auto t2 = Clock::now();

    auto duration = chrono::duration_cast<chrono::microseconds>(t2 - t1).count();

    cout << name
         << " | Tiempo: " << duration << " μs"
         << " | Comparaciones: " << m.comparisons
         << " | Intercambios: " << m.swaps << "\n";
}

// ======================= MAIN =======================
int main() {
    srand(time(0));

    int n = 1000;

    vector<int> randomData(n);
    for (int i = 0; i < n; i++) randomData[i] = rand() % 10000;

    vector<int> sortedData = randomData;
    sort(sortedData.begin(), sortedData.end());

    vector<int> reverseData = sortedData;
    reverse(reverseData.begin(), reverseData.end());

    cout << "===== DATOS ALEATORIOS =====\n";
    test(randomData, bubbleSort,   "Bubble Sort   ");
    test(randomData, insertionSort,"Insertion Sort");
    test(randomData, mergeSort,    "Merge Sort    ");
    test(randomData, quickSort,    "Quick Sort    ");

    cout << "\n===== DATOS ORDENADOS =====\n";
    test(sortedData, bubbleSort,   "Bubble Sort   ");
    test(sortedData, insertionSort,"Insertion Sort");
    test(sortedData, mergeSort,    "Merge Sort    ");
    test(sortedData, quickSort,    "Quick Sort    ");

    cout << "\n===== DATOS INVERSOS =====\n";
    test(reverseData, bubbleSort,   "Bubble Sort   ");
    test(reverseData, insertionSort,"Insertion Sort");
    test(reverseData, mergeSort,    "Merge Sort    ");
    test(reverseData, quickSort,    "Quick Sort    ");

    return 0;
}

File: unidad-1/prac5-6/algoritmos.py
import random
import time

# ======================= MÉTRICAS =======================
class Metrics:
    def __init__(self):
        self.comparisons = 0
        self.swaps = 0

# ======================= BUBBLE SORT =======================
def bubble_sort(arr, m):
    n = len(arr)
    for i in range(n):
        swapped = False
        for j in range(0, n - i - 1):
            m.comparisons += 1
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                m.swaps += 1
                swapped = True
        if not swapped:
            break

# ======================= INSERTION SORT =======================
def insertion_sort(arr, m):
    for i in range(1, len(arr)):
        key = arr[i]
        j = i - 1
        while j >= 0:
            m.comparisons += 1
            if arr[j] > key:
                arr[j + 1] = arr[j]
                m.swaps += 1
                j -= 1
            else:
                break
        arr[j + 1] = key

# ======================= MERGE SORT =======================
def merge(arr, left, mid, right, m):
    L = arr[left:mid+1]
    R = arr[mid+1:right+1]

    i = j = 0
    k = left

    while i < len(L) and j < len(R):
        m.comparisons += 1
        if L[i] <= R[j]:
            arr[k] = L[i]
            i += 1
        else:
            arr[k] = R[j]
            j += 1
        m.swaps += 1
        k += 1

    while i < len(L):
        arr[k] = L[i]
        i += 1
        k += 1
        m.swaps += 1

    while j < len(R):
        arr[k] = R[j]
        j += 1
        k += 1
        m.swaps += 1

def merge_sort_rec(arr, left, right, m):
    if left < right:
        mid = (left + right) // 2
        merge_sort_rec(arr, left, mid, m)
        merge_sort_rec(arr, mid + 1, right, m)
        merge(arr, left, mid, right, m)

def merge_sort(arr, m):
    merge_sort_rec(arr, 0, len(arr) - 1, m)

# ======================= QUICK SORT =======================
def partition(arr, low, high, m):
    pivot = arr[high]
    i = low - 1

    for j in range(low, high):
        m.comparisons += 1
        if arr[j] < pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
            m.swaps += 1

    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    m.swaps += 1
    return i + 1

def quick_sort_rec(arr, low, high, m):
    if low < high:
        pi = partition(arr, low, high, m)
        quick_sort_rec(arr, low, pi - 1, m)
        quick_sort_rec(arr, pi + 1, high, m)

def quick_sort(arr, m):
    quick_sort_rec(arr, 0, len(arr) - 1, m)

# ======================= TEST =======================
def test(data, func, name):
    arr = data.copy()
    m = Metrics()

    start = time.time()
    func(arr, m)
    end = time.time()

    print(f"{name} | Tiempo: {(end - start)*1e6:.2f} μs"
          f" | Comparaciones: {m.comparisons}"
          f" | Intercambios: {m.swaps}")

# ======================= MAIN =======================
n = 1000

# Datos aleatorios
random_data = [random.randint(0, 10000) for _ in range(n)]

# Datos ordenados
sorted_data = sorted(random_data)

# Datos inversos
reverse_data = sorted_data[::-1]

print("===== DATOS ALEATORIOS =====")
test(random_data, bubble_sort,    "Bubble Sort   ")
test(random_data, insertion_sort, "Insertion Sort")
test(random_data, merge_sort,     "Merge Sort    ")
test(random_data, quick_sort,     "Quick Sort    ")

print("\n===== DATOS ORDENADOS =====")
test(sorted_data, bubble_sort,    "Bubble Sort   ")
test(sorted_data, insertion_sort, "Insertion Sort")
test(sorted_data, merge_sort,     "Merge Sort    ")
test(sorted_data, quick_sort,     "Quick Sort    ")

print("\n===== DATOS INVERSOS =====")
test(reverse_data, bubble_sort,    "Bubble Sort   ")
test(reverse_data, insertion_sort, "Insertion Sort")
test(reverse_data, merge_sort,     "Merge Sort    ")
test(reverse_data, quick_sort,     "Quick Sort    ")